### ISSUE TITLE
Mixtral branch : What happens when we give both positive and negative prompts per an expert ? 

### ISSUE THREAD

**Issue by shamanez** [cg123/mergekit#64]
As far as I understand in this [line](https://github.com/cg123/mergekit/blob/mixtral/mergekit/scripts/mixtral_moe.py#L137), it seems like it only uses either only negative prompts even if we give both negative and positive prompts. 

@cg123  @DocShotgun  @q5sys

**Comment by cg123** [comment]
My reasoning for that was pretty straightforward. We want the gate vector to maximize its dot product with the positive hidden states and minimize the dot product with the negative hidden states.

I use these two properties of the dot product here:
* Given a set of unit vectors, the unit vector that maximizes the sum of dot products with the set is the (normalized) average of the set
* Minimizing the dot product with a vector is equivalent to maximizing the dot product with the vector negated

I think I got things right there but if I bungled the math definitely let me know. :)

**Comment by shamanez** [comment]
Yeah, it seems ok. But I'll definitely think of it. It is more like getting a good representation for each expert.

**Comment by win10ogod** [comment]
@cg123 
Can we use algorithms to automatically optimize the merging of model weights and the creation of expert routers along the most efficient path?

**Comment by shamanez** [comment]
@win10ogod If we gather sufficient positive and negative pairs, we could employ techniques like triplet loss or in-batch negative contrastive loss (which doesn't require negative examples) to train a compact sentence transformer. This would enable us to create individual embeddings for each router.
